libcudacxx/include/cuda/__iterator/permutation_iterator.h

```diff
@@ -0,0 +1,424 @@
+
+  //! @brief Compares two \c permutation_iterator for less than, by comparing the memory location they point at
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _OtherIter, class _OtherOffset)
+  _CCCL_REQUIRES(_CUDA_VSTD::totally_ordered_with<_Offset, _OtherOffset>)
+  [[nodiscard]] _CCCL_API friend constexpr bool operator<(
+    const permutation_iterator& __lhs,
+    const permutation_iterator<_OtherIter, _OtherOffset>& __rhs) noexcept(__nothrow_less_than<_Offset, _OtherOffset>)
+  {
+    return __lhs.__offset_ < __rhs.offset();
+  }
+
+  //! @brief Compares two \c permutation_iterator for less equal, by comparing the memory location they point at
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _OtherIter, class _OtherOffset)
+  _CCCL_REQUIRES(_CUDA_VSTD::totally_ordered_with<_Offset, _OtherOffset>)
+  [[nodiscard]] _CCCL_API friend constexpr bool operator<=(
+    const permutation_iterator& __lhs,
+    const permutation_iterator<_OtherIter, _OtherOffset>& __rhs) noexcept(__nothrow_less_equal<_Offset, _OtherOffset>)
+  {
+    return __lhs.__offset_ <= __rhs.offset();
+  }
+
+  //! @brief Compares two \c permutation_iterator for greater than, by comparing the memory location they point at
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _OtherIter, class _OtherOffset)
+  _CCCL_REQUIRES(_CUDA_VSTD::totally_ordered_with<_Offset, _OtherOffset>)
+  [[nodiscard]] _CCCL_API friend constexpr bool operator>(
+    const permutation_iterator& __lhs,
+    const permutation_iterator<_OtherIter, _OtherOffset>& __rhs) noexcept(__nothrow_greater_than<_Offset, _OtherOffset>)
+  {
+    return __lhs.__offset_ > __rhs.offset();
+  }
+
+  //! @brief Compares two \c permutation_iterator for greater equal, by comparing the memory location they point at
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _OtherIter, class _OtherOffset)
+  _CCCL_REQUIRES(_CUDA_VSTD::totally_ordered_with<_Offset, _OtherOffset>)
+  [[nodiscard]] _CCCL_API friend constexpr bool
+  operator>=(const permutation_iterator& __lhs, const permutation_iterator<_OtherIter, _OtherOffset>& __rhs) noexcept(
+    __nothrow_greater_equal<_Offset, _OtherOffset>)
+  {
+    return __lhs.__offset_ >= __rhs.offset();
+  }
+};
+
+_CCCL_TEMPLATE(class _Iter, class _Offset)
+_CCCL_REQUIRES(_CUDA_VSTD::random_access_iterator<_Iter> _CCCL_AND _CUDA_VSTD::random_access_iterator<_Offset>)
+_CCCL_HOST_DEVICE permutation_iterator(_Iter, _Offset) -> permutation_iterator<_Iter, _Offset>;
+
+//! @brief Creates an \c permutation_iterator from an iterator and an iterator to an integral offset
+//! @param __iter The iterator
+//! @param __offset The iterator to an integral offset
+_CCCL_TEMPLATE(class _Iter, class _Offset)
+_CCCL_REQUIRES(_CUDA_VSTD::random_access_iterator<_Iter> _CCCL_AND _CUDA_VSTD::random_access_iterator<_Offset>)
+[[nodiscard]] _CCCL_API constexpr permutation_iterator<_Iter, _Offset>
+make_permutation_iterator(_Iter __iter, _Offset __offset) noexcept(
```

Comment:
Does this also just exist for migrating from Thrust or is there some reason to use this over CTAD?

Reply:
Yes, our users extensively use the `make_meow_iterator` functions.